### PR TITLE
[Eval] Unique samples in hashtableSampleEntries

### PR DESCRIPTION
**Pull request by sarthakaggarwal97**
Mirror of valkey-io/valkey#3460 for CI agent eval. Do not merge.

### CODE REVIEW

**Comment by sarthakaggarwal97** [comment]
## PR Summary

The PR splits what used to be a single sampler into two: `hashtableSampleEntries` now scans contiguous buckets from a random starting cursor (so results are unique), and a new `sampleRandomBuckets` pi

### Walkthrough

The PR splits what used to be a single sampler into two: `hashtableSampleEntries` now scans contiguous buckets from a random starting cursor (so results are unique), and a new `sampleRandomBuckets` picks an independent random bucket per scan for fair random-entry selection at the cost of possible duplicates. `hashtableRandomEntry` and `hashtableFairRandomEntry` move to the new function; other callers (t_hash, t_set, t_zset, kvstore, cluster gossip) keep the now-unique-sampling variant.

The motivation is a flaky cluster gossip test and a real behavior regression: with the previous implementation each scan call restarted at a new random cursor, so `hashtableSampleEntries` could return duplicate entries even when the table had plenty of candidates. That depressed gossip counts on the wire and broke callers that assumed N distinct entries. Restoring contiguous scan order matches the old `dictGetSomeKeys` shape and lets `packet.tcl` tighten its assertion from `> 3` to `[7, 9]`.

A couple of things worth a look. The tightened gossip count range is the main flake risk — if 7 isn't a hard floor from the math, CI will rediscover it; some justification in the commit message would help. The empty-table infinite-loop fix (2cc5bc6b) restores a guard that was dropped earlier in the branch, which is a little worrying — a quick unit test for `hashtableRandomEntry` / `sampleRandomBuckets` on an empty and near-empty table would prevent that class of regression from recurring. Minor: in `sampleRandomBuckets` the `samples.seen <= count ? samples.seen : count` return is effectively always `count` given the loop condition and the initial cap, so it could just be `return count;`. DCO trailers look fine on all non-merge commits.

### Maintainer Checklist

- DCO: commit(s) missing `Signed-off-by` trailers: `cb8d155eaaaa`, `5fb5fec8c088`, `757d4dc94b6a`.
- Extra tests: this change touches riskier Valkey areas for `unstable`; consider the `run-extra-tests` label.
- Labels: consider applying `pending-missing-dco`, `run-extra-tests`.

File: src/hashtable.c [hunk 1]
```diff
@@ -2325,11 +2325,28 @@ bool hashtableNext(hashtableIterator *iterator, void **elemptr) {
...
+    scan_samples samples;
+    samples.size = count;
+    samples.seen = 0;
+    samples.entries = dst;
+    while (samples.seen < count) {
+        hashtableScan(ht, randomSizeT(), sampleEntriesScanFn, &samples);
+    }
+    rehashStepOnReadIfNeeded(ht);
+    return samples.seen <= count ? samples.seen : count;
+}
```
Comment (sarthakaggarwal97) [inline src/hashtable.c:2343]: `sampleRandomBuckets` is new and is the code path exercised by `hashtableRandomEntry` / `hashtableFairRandomEntry`. The existing `test_hashtable.cpp` uses `hashtableFairRandomEntry` but only asserts a non-null entry — it doesn't verify the fairness/duplicate behavior this commit is introducing, nor the empty-table and sparse-table corner cases that motivated the follow-up `hashtableSize` guard. A small unit test that at least exercises the empty/size-< count paths would protect against regressions like the one fixed in 2cc5bc6b.

File: tests/unit/cluster/packet.tcl [hunk 2]
```diff
@@ -134,7 +134,7 @@ start_cluster 10 0 {tags {external:skip cluster tls:skip}} {
         binary scan $reply @14Su count
 
         assert_equal 1 $type
-        assert_morethan $count 3
+        assert_morethan_equal $count 7
```
Comment (sarthakaggarwal97) [inline tests/unit/cluster/packet.tcl:137]: With 10 nodes and `cluster-message-gossip-perc=80`, `wanted = min(10*80/100, freshnodes) = min(8, 8) = 8`, and the gossip loop exits at `gossipcount >= wanted`. So in this test, count can never exceed 8 (no PFAIL nodes). Not a bug — the `<= 9` has slack in it — but since the PR is specifically tightening the assertion bounds to reflect the new uniqueness guarantee, you may want to tighten the upper bound to 8 as well for symmetry.

**Review by sarthakaggarwal97 — COMMENTED** [review]
Automated review found 2 issue(s). Highest severity: `low`.
